zebra-state/src/service/read/find.rs

```diff
@@ -526,3 +535,102 @@ where
 
     collect_chain_headers(chain, db, intersection, stop, max_len)
 }
+
+/// Returns the median-time-past of the *next* block to be added to the best chain in
+/// `non_finalized_state` or `db`.
+///
+/// # Panics
+///
+/// - If we don't have enough blocks in the state.
+pub fn next_median_time_past(
+    non_finalized_state: &NonFinalizedState,
+    db: &ZebraDb,
+) -> Result<DateTime32, BoxError> {
+    let mut best_relevant_chain_result = best_relevant_chain(non_finalized_state, db);
+
+    // Retry the finalized state query if it was interrupted by a finalizing block.
+    //
+    // TODO: refactor this into a generic retry(finalized_closure, process_and_check_closure) fn
+    for _ in 0..FINALIZED_STATE_QUERY_RETRIES {
+        if best_relevant_chain_result.is_ok() {
+            break;
+        }
+
+        best_relevant_chain_result = best_relevant_chain(non_finalized_state, db);
+    }
+
+    Ok(calculate_median_time_past(best_relevant_chain_result?))
+}
+
+/// Do a consistency check by checking the finalized tip before and after all other database queries.
+///
+/// Returns recent blocks in reverse height order from the tip.
+/// Returns an error if the tip obtained before and after is not the same.
+///
+/// # Panics
+///
+/// - If we don't have enough blocks in the state.
+fn best_relevant_chain(
+    non_finalized_state: &NonFinalizedState,
+    db: &ZebraDb,
+) -> Result<[Arc<Block>; POW_ADJUSTMENT_BLOCK_SPAN], BoxError> {
+    let state_tip_before_queries = read::best_tip(non_finalized_state, db).ok_or_else(|| {
+        BoxError::from("Zebra's state is empty, wait until it syncs to the chain tip")
+    })?;
+
+    let best_relevant_chain =
+        any_ancestor_blocks(non_finalized_state, db, state_tip_before_queries.1);
+    let best_relevant_chain: Vec<_> = best_relevant_chain
+        .into_iter()
+        .take(POW_ADJUSTMENT_BLOCK_SPAN)
+        .collect();
+    let best_relevant_chain = best_relevant_chain.try_into().map_err(|_error| {
+        "Zebra's state only has a few blocks, wait until it syncs to the chain tip"
+    })?;
+
+    let state_tip_after_queries =
+        read::best_tip(non_finalized_state, db).expect("already checked for an empty tip");
+
+    if state_tip_before_queries != state_tip_after_queries {
+        return Err("Zebra is committing too many blocks to the state, \
+                    wait until it syncs to the chain tip"
+            .into());
+    }
```

Comment:
That's pretty tricky, because we don't need the history tree when we're just getting the median-time-past.

So I think the simplest way to do this is to refactor all the retry functions to use a generic "retry" closure. I've left TODOs, but I don't want to do that refactor in this PR. And it will change a bunch of existing production code, so we might want to wait until after the audit.